# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def add_two(x):
    return x + 2

print(add_two)
add_two

<function add_two at 0x000002084F44B7F0>


<function __main__.add_two(x)>

In [2]:
# lambda param: one-line function
lambda x: x + 2

<function __main__.<lambda>(x)>

In [3]:
print(lambda x: x + 2)

<function <lambda> at 0x000002084F5789D0>


In [4]:
add_two(10)

12

In [6]:
# IIFE - Immediately Invoked Function Expression
# Execute a lambda function immediately after defining the function
# Syntax: (lambda param: expression)(arg)

(lambda x: x + 2)(10)

12

In [7]:
x = 10

y = (lambda num: num**2)(x)

print(y)

100


#### Saving to a Variable

In [8]:
named_lambda = lambda z: z**3

print(named_lambda)

<function <lambda> at 0x000002084F5789D0>


In [9]:
named_lambda(5)

125

#### Multiple Inputs

In [14]:
# Syntax: lambda param1, param2, etc: expression

def multiply(x, y, z):
    return x * y * z

multiply(10, 4, 2)

80

In [15]:
multiply(2, 7, 3)

42

In [16]:
# As an IIFE - (lambda param1, param2, etc: expression)(arg1, arg2, etc.)
(lambda x, y, z: x * y * z)(11, 5, 2)

110

In [17]:
(lambda x, y, z: x * y * z)(10, 4, 3)

120

In [20]:
def accept_args(*args):
    return args

accept_args('a', 'b', 'c')

('a', 'b', 'c')

In [28]:
(lambda *args: args)('a', 'b', 'c')

('a', 'b', 'c')

In [24]:
def accept_kwargs(**kwargs):
    return kwargs

accept_kwargs(name='Brian', city='Chicago', age=55)

{'name': 'Brian', 'city': 'Chicago', 'age': 55}

In [32]:
(lambda **kwargs: kwargs)(name='Sarah', city='Canon City', age=99)

{'name': 'Sarah', 'city': 'Canon City', 'age': 99}

#### Passing a Function into a Function

In [33]:
def apply_to_each(f, a_list):
    """
    Print the reutrn value of function f applied to every element in a_list
    f -> Function to execute
    a_list -> List of elements to execute function f on
    """
    for element in a_list:
        f_return = f(element)
        print(f_return)

In [36]:
def get_first_letter(string):
    return string[0]


In [37]:
list_of_seasons = ['Winter', 'Spring', 'Summer', 'Fall']

In [39]:
apply_to_each(get_first_letter, list_of_seasons)

W
S
S
F


In [41]:
def reverse_string(string):
    return string[::-1]


In [42]:
apply_to_each(reverse_string, list_of_seasons)

retniW
gnirpS
remmuS
llaF


#### Passing a Lambda into a Function

In [45]:
# Call the apply_to_each function, pass in a lambda function as f argument

apply_to_each(lambda element: element[::-1], list_of_seasons)

retniW
gnirpS
remmuS
llaF


In [47]:
apply_to_each(lambda element: element[::-1], [[1, 2, 3], [4, 5, 6]])

[3, 2, 1]
[6, 5, 4]


In [48]:
def square_num(num):
    return num ** 2

apply_to_each(square_num, [1, 2, 3, 4, 5])

1
4
9
16
25


In [49]:
apply_to_each(lambda num: num**2, [1, 2, 3, 4, 5])

1
4
9
16
25


#### Returning a Function from a Function

In [59]:
def my_outer_function(name):
    print(f"Hello {name}")
    
    # Define a function - will be scoped to the function
    def my_inner_function():
        city = 'Chicago'
        print(f'This is the inner function, {name} from {city}')
        
    # Execute the inner function 
    my_inner_function()
    
#     print(f"Goodbye {name} from {city}") # NameError: name 'city' is not defined
    
    
print(my_outer_function)
my_outer_function

<function my_outer_function at 0x000002084F4C6830>


<function __main__.my_outer_function(name)>

In [60]:
my_outer_function('Brian')

Hello Brian
This is the inner function, Brian from Chicago


In [62]:
my_outer_return = my_outer_function('Sarah')
print(my_outer_return)

Hello Sarah
This is the inner function, Sarah from Chicago
None


In [61]:
def my_outer_function2(name):
    print(f"Hello {name}")
    
    # Define a function - will be scoped to the function
    def my_inner_function():
        city = 'Chicago'
        print(f'This is the inner function, {name} from {city}')
        
    # return the inner function 
    return my_inner_function

print(my_outer_function2)
my_outer_function2

<function my_outer_function2 at 0x0000020851071990>


<function __main__.my_outer_function2(name)>

In [63]:
my_outer_return2 = my_outer_function2('Brian')
print(my_outer_return2)

Hello Brian
<function my_outer_function2.<locals>.my_inner_function at 0x0000020851073F40>


In [65]:
my_outer_return2()

This is the inner function, Brian from Chicago


In [68]:
def create_multiplier(num):
    def multiply(x):
        return num * x
    return multiply

five_times = create_multiplier(5)
print(five_times)

<function create_multiplier.<locals>.multiply at 0x0000020851131000>


In [69]:
five_times(10)

50

In [70]:
five_times(6)

30

In [71]:
three_times = create_multiplier(3)
print(three_times)

<function create_multiplier.<locals>.multiply at 0x0000020851070DC0>


In [72]:
three_times(10)

30

In [73]:
three_times(6)

18

#### Returning a Lambda from a Function

In [76]:
def new_multiplier(num):
    return lambda x: num * x


four_times = new_multiplier(4)
print(four_times)

<function new_multiplier.<locals>.<lambda> at 0x0000020851073A30>


In [77]:
four_times(10)

40

In [78]:
four_times(5)

20

#### If Statements within Lambdas

In [79]:
def double_pos_square_neg(number):
    if number > 0:
        return number * 2
    else:
        return number ** 2
    
    
print(double_pos_square_neg(10))
print(double_pos_square_neg(20))
print(double_pos_square_neg(-10))
print(double_pos_square_neg(-20))

20
40
100
400


In [80]:
# Ternary: value_if_true if condition else value_if_false
def double_pos_square_neg(number):
#     if number > 0:
#         return number * 2
#     else:
#         return number ** 2
    return number * 2 if number > 0 else number ** 2
    
    
print(double_pos_square_neg(10))
print(double_pos_square_neg(20))
print(double_pos_square_neg(-10))
print(double_pos_square_neg(-20))

20
40
100
400


In [81]:
# As an IIFE
(lambda num: num * 2 if num > 0 else num ** 2)(10)

20

In [82]:
(lambda num: num * 2 if num > 0 else num ** 2)(-10)

100

In [84]:
# Passing it into a function
my_nums = [i for i in range(-5, 6)]
print(my_nums)
apply_to_each(lambda num: num * 2 if num > 0 else num ** 2, my_nums)

[-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5]
25
16
9
4
1
0
2
4
6
8
10


#### In-Class Exercise #1 <br>
<p>Write a lambda function that takes in three arguments <code>a</code>,<code>b</code>,<code>c</code> and returns the product of <code>a</code> and the difference of <code>b</code> and <code>c</code></p>

In [ ]:
# Example 1 
# a = 10, b = 5, c = 2
# Output: 30

# Example 2
# a = 20, b = 10, c = 5
# Output: 100

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

In [85]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [88]:
# Function that we want to apply to each element in an iterable
def squared_minus_two(num):
    return num ** 2 - 2

numbers = [3, 5, 7, 8, 2, 1, 4, 2, 8]

# Create an empty list
new_nums = []
# Loop over the iterable (numbers)
for num in numbers:
    # Apply the function to each element
    new_num = squared_minus_two(num)
    # Add the new num to the new_nums list
    new_nums.append(new_num)
    
print(new_nums)

[7, 23, 47, 62, 2, -1, 14, 2, 62]


In [90]:
# map(function, *iterables -> List, Tuple, String, Set, Dictionary, etc)
list(map(squared_minus_two, numbers))

[7, 23, 47, 62, 2, -1, 14, 2, 62]

In [108]:
my_map_obj = map(squared_minus_two, numbers)
print(my_map_obj)

In [109]:
help(next)

Help on built-in function next in module builtins:

next(...)
    next(iterator[, default])
    
    Return the next item from the iterator. If default is given and the iterator
    is exhausted, it is returned instead of raising StopIteration.



In [110]:
next(my_map_obj)

7

In [111]:
for num in my_map_obj:
    print(num)

23
47
62
2
-1
14
2
62


In [116]:
x_values = [1, 2, 3, 4, 5]
y_values = [10, 20, 30, 40, 50]

def multiply(x, y):
    print(f"x={x}, y={y}")
    return x * y

list(map(multiply, x_values, y_values))

x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50


[10, 40, 90, 160, 250]

In [125]:
x_values2 = [1, 2, 3, 4, 5]
y_values2 = [10, 20, 30, 40, 50, 60, 70, 80]

nums_mult = list(map(multiply, x_values2, y_values2))

print(nums_mult)

x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50
[10, 40, 90, 160, 250]


In [126]:
nums_mult = []
# shorter = len(x_values2) if len(x_values2) < len(y_values2) else len(y_values2)
for i in range(len(x_values2)):
    nums_mult.append(multiply(x_values2[i], y_values2[i]))
print(nums_mult)

x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50
[10, 40, 90, 160, 250]


#### Using Lambda's with Map

In [127]:
nums_again = [4, 7, 2, 4, 7, 8, 2, 1, 9]

new_nums = list(map(lambda t: t**2 -2 , nums_again))
print(new_nums)

[14, 47, 2, 14, 47, 62, 2, -1, 79]


In [128]:
new2 = list(map(lambda x: x**3, [1, 2, 3, 4, 5]))
new2

[1, 8, 27, 64, 125]

In [130]:
x_vals = [5, 10, 15, 20, 25]
y_vals = [10, 20, 30, 40, 50]

x_y_vals = list(map(lambda a,b: a+b, x_vals, y_vals))
print(x_y_vals)

[15, 30, 45, 60, 75]


In [132]:
alphabet = 'abcdefghijklmnopqrstuvwxyz'

unicode = list(map(lambda letter: ord(letter), alphabet))

print(unicode)

[97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122]


In [134]:
vowels = ('a', 'e', 'i', 'o', 'u')

upper_vowels = tuple(map(lambda v: v.upper(), vowels))
print(upper_vowels)

('A', 'E', 'I', 'O', 'U')


In [135]:
squares = list(map(lambda x: x**2, range(10,20)))
squares

[100, 121, 144, 169, 196, 225, 256, 289, 324, 361]

In [136]:
hellos = list(map(lambda x: 'Hello World', [1, 2, 3, 4, 5]))
hellos

['Hello World', 'Hello World', 'Hello World', 'Hello World', 'Hello World']

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number if it is even and halve it if it is odd</p>

In [ ]:
numbers = range(1,11)

# [0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]



## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

In [137]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [138]:
names = ['Bob', 'Billy', 'Alex', 'Barbara', 'Amy', 'Andy', 'Catherine', 'Cameron']

def starts_with_a(name):
    return name[0].upper() == 'A'

a_names_map = list(map(starts_with_a, names))
print(a_names_map)

[False, False, True, False, True, True, False, False]


In [139]:
# Syntax: filter(function_to_apply, iterable -> List,Tuple,Set,Dict,String,etc)

a_names_filter = list(filter(starts_with_a, names))
print(a_names_filter)

['Alex', 'Amy', 'Andy']


In [140]:
def remainder(num):
    # Takes in a num and returns the remainder after dividing by 3
    return num % 3

my_nums = [x for x in range(1,11)]
print(my_nums)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [141]:
list(map(remainder, my_nums))

[1, 2, 0, 1, 2, 0, 1, 2, 0, 1]

In [142]:
list(filter(remainder, my_nums))

[1, 2, 4, 5, 7, 8, 10]

In [143]:
for i in range(-5, 6):
    print(i, bool(i))

-5 True
-4 True
-3 True
-2 True
-1 True
0 False
1 True
2 True
3 True
4 True
5 True


#### Using Lambda's with Filter()

In [147]:
print(names)

['Bob', 'Billy', 'Alex', 'Barbara', 'Amy', 'Andy', 'Catherine', 'Cameron']


In [145]:
b_names = list(filter(lambda name: True if name[0].upper() == 'B' else False, names))
print(b_names)

['Bob', 'Billy', 'Barbara']


In [146]:
list(filter(lambda name: name[0].upper() == 'C', names))

['Catherine', 'Cameron']

In [148]:
def get_c_names(list_of_names):
    return list(filter(lambda n: n[0].upper() == 'C', list_of_names))


print(get_c_names(names))
print(get_c_names(['Cindy', 'Margaret', 'Michael', 'Chelsea', 'Jack', 'Connor', 'Colin']))

['Catherine', 'Cameron']
['Cindy', 'Chelsea', 'Connor', 'Colin']


#### In-Class Exercise #3 <br>
<p>Write a function that takes in a list of cities as strings and returns a list of cities with at least two words in the name</p>

In [152]:
cities = ['Chicago', 'Los Angeles', 'New York City', 'Miami', 'Oklahoma City', 'Baton Rouge', 'Boston', 'Denver']
# Output: ['Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge'] 
two_words = list(filter(lambda city: len(city.split()) > 1, cities))
print(two_words)

['Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge']


In [159]:
def is_two_words(city):
    return len(city.split()) >= 2


def cities_with_2_words(func, cities):
    cities2 = []
    for city in cities:
        if func(city):
            cities2.append(city)
    print(cities2)        
            
    
cities_with_2_words(is_two_words, cities)

['Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge']


In [161]:
cities_with_2_words(is_two_words, ['New York', 'Atlanta', 'Santa Fe', 'San Francisco'])

['New York', 'Santa Fe', 'San Francisco']


In [162]:
cities_with_2_words(lambda city: len(city.split()) >= 2, ['San Jose', 'Inidanapolis', 'Santa Barbara'])

['San Jose', 'Santa Barbara']


In [164]:
list(filter(lambda city: len(city.split()) >= 2, ['San Jose', 'Inidanapolis', 'Santa Barbara']))

['San Jose', 'Santa Barbara']

In [165]:
def filter_func(function_to_test, a_list):
    new_list = []
    for item in a_list:
        if function_to_test(item):
            new_list.append(item)
    return new_list


filter_func(lambda x: x % 2 == 0, [1, 2, 3, 4, 5, 6, 7, 8])

[2, 4, 6, 8]

In [166]:
import re

In [168]:
two_word_plus_pattern = re.compile('([A-Z][A-z]+) ([A-Z][A-z]+\s?)+')

list(filter(lambda city: two_word_plus_pattern.match(city), cities))

['Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge']

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [170]:
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, iterable[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence
    or iterable, from left to right, so as to reduce the iterable to a single
    value.  For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the iterable in the calculation, and serves as a default when the
    iterable is empty.



In [172]:
def add_nums(x, y):
    print('='*20)
    print(f"x: {x}")
    print(f"y: {y}")
    print(f"returns: {x + y}")
    return x + y

result_add = reduce(add_nums, [1, 2, 3, 4, 5])
print(result_add)

x: 1
y: 2
returns: 3
x: 3
y: 3
returns: 6
x: 6
y: 4
returns: 10
x: 10
y: 5
returns: 15
15


In [173]:
total = 0
for num in [1, 2, 3, 4, 5]:
    total += num
    
print(total)

15


#### Using Lambda's with Reduce()

In [174]:
lambda_reduce = reduce(lambda a,b: a + b, [10, 20, 30, 40, 50])
print(lambda_reduce)

150


In [175]:
max_value = reduce(lambda x, y: x if x > y else y, [10, 5, 8, 12, 9, 4])
print(max_value)

12


In [179]:
from random import randint
randoms = [randint(1, 50) for _ in range(20)]
print(randoms)

[27, 36, 14, 28, 45, 22, 38, 49, 23, 31, 36, 39, 11, 28, 12, 3, 10, 6, 12, 15]


In [181]:
def get_larger(x, y):
    print('='*20)
    print(f"x: {x}")
    print(f"y: {y}")
    print(f"returns: {x if x > y else y}")
    return x if x > y else y


max_value = reduce(get_larger, randoms)
print(max_value)

x: 27
y: 36
returns: 36
x: 36
y: 14
returns: 36
x: 36
y: 28
returns: 36
x: 36
y: 45
returns: 45
x: 45
y: 22
returns: 45
x: 45
y: 38
returns: 45
x: 45
y: 49
returns: 49
x: 49
y: 23
returns: 49
x: 49
y: 31
returns: 49
x: 49
y: 36
returns: 49
x: 49
y: 39
returns: 49
x: 49
y: 11
returns: 49
x: 49
y: 28
returns: 49
x: 49
y: 12
returns: 49
x: 49
y: 3
returns: 49
x: 49
y: 10
returns: 49
x: 49
y: 6
returns: 49
x: 49
y: 12
returns: 49
x: 49
y: 15
returns: 49
49


In [184]:
print(randoms)

[27, 36, 14, 28, 45, 22, 38, 49, 23, 31, 36, 39, 11, 28, 12, 3, 10, 6, 12, 15]


In [182]:
current_max = randoms[0]

for num in randoms:
    if num > current_max:
        current_max = num
print(current_max)

49


In [183]:
max(randoms)

49

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [189]:
nums = [2, 4, 6, 8, 10]
# Output: 3840
reduce(lambda num1, num2: num1 * num2, nums)

3840

In [188]:
prod = 1
for num in nums:
    prod *= num
    
print(prod)

3840


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [190]:
def capitalize(name):
    # Takes in a name, upper case the first letter and lower case the rest
    return name[0].upper() + name[1:].lower()

def full_name(first, last):
    # take in a first name, capitalize the first name, take in a last name, capitalize the last name and 
    # then returns the concatenated string of first + space + last
    return capitalize(first) + ' ' + capitalize(last)


def greet_person(first_name, last_name, day_of_week):
    person_full_name = full_name(first_name, last_name)
    print(f"Hello {person_full_name}, it sure is beautiful on this {capitalize(day_of_week)}")
    
greet_person('bRIaN', 'stANtOn', 'tUEsdaY')

Hello Brian Stanton, it sure is beautiful on this Tuesday


In [193]:
import time

def func_one():
    print("Function One Started")
    time.sleep(5)
    print("Function One Ended")
    
    
def func_two():
    print("Function Two Started")
    func_one()
    print("Function Two Ended")
    
    
func_two()

Function Two Started
Function One Started
Function One Ended
Function Two Ended


#### Implementing a Base Case

In [1]:
def add_nums(num):
    if num <= 1:
        return num
    else:
        return num + add_nums(num - 1)
    
    
add_nums(10)

55

In [3]:
def add_nums(num):
    # Set up a base case
    if num <= 1:
        print(f"add_nums({num}) = {num}")
        return num
    else:
        print(f"add_nums({num}) = {num} + add_nums({num - 1})")
        return num + add_nums(num - 1)
    
    
add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

In [2]:
# Can also be used as a way of looping
def print_recursion(a_list):
    if a_list:
        print(a_list[-1])
        print_recursion(a_list[:-1])
        
        
        
print_recursion(['red', 'orange', 'yellow', 'green', 'blue', 'indigo'])

indigo
blue
green
yellow
orange
red


#### Writing a Factorial Function

In [3]:
# 5! = 5 * 4!
# 4! = 4 * 3!
# ...
# 1! = 1

def factorial(num):
    if num <= 1:
        return num
    else:
        return num * factorial(num - 1)
    
factorial(5)

120

In [40]:
def recursive_example(number):
        if number <= 3:
            return number
        else:
            return recursive_example(number - 1) * recursive_example(number - 2) / recursive_example(number - 3)

recursive_example(30)

9565938.0

In [42]:
def hide_the_cache():
    cache = {}
    
    def recursive_example(number):
        if number in cache:
            return cache[number]
        elif number <= 3:
            return number
        else:
            answer = recursive_example(number - 1) * recursive_example(number - 2) / recursive_example(number - 3)
            cache[number] = answer
            return answer
    return recursive_example
    
    
recursive_func = hide_the_cache()
recursive_func(60)

137260754729766.05

In [50]:
def hide_the_cache():
    cache = {}

    def recursive_example(number):
        if number in cache:
            return cache[number]
        elif number <= 3:
            return number
        else:
            answer = recursive_example(number - 1) * recursive_example(number - 2) / recursive_example(number - 3)
            cache[number] = answer
            return answer
    
    return recursive_example
    
my_func = hide_the_cache()
print(my_func)

<function hide_the_cache.<locals>.recursive_example at 0x000002BAB301A320>


In [51]:
my_func(100)

4.78598658461241e+23

In [52]:
cache

{}

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [53]:
lengths = map(len, ['Brian', 'Stanton', 'Coding Temple', 'Instructor'])

print(lengths)

In [55]:
for item in lengths:
    print(item)

5
7
13
10


In [56]:
for item in lengths:
    print(item)

In [58]:
help(iter)

Help on built-in function iter in module builtins:

iter(...)
    iter(iterable) -> iterator
    iter(callable, sentinel) -> iterator
    
    Get an iterator from an object.  In the first form, the argument must
    supply its own iterator, or be a sequence.
    In the second form, the callable is called until it returns the sentinel.



In [57]:
help(next)

Help on built-in function next in module builtins:

next(...)
    next(iterator[, default])
    
    Return the next item from the iterator. If default is given and the iterator
    is exhausted, it is returned instead of raising StopIteration.



In [60]:
a_new_list = ['hi', 'hello', 'how are you', 'goodbye']

list_iter = iter(a_new_list)
print(list_iter)

In [65]:
next(list_iter)

StopIteration: 

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [68]:
def my_range_iter(start, stop, step=1):
    while start < stop:
        return start
        start += step
        
        
my_gen = my_range_iter(0, 10)
print(my_gen)

0


In [69]:
def my_range_iter(start, stop, step=1):
    while start < stop:
        yield start
        start += step
        
        
my_gen = my_range_iter(0, 10)
print(my_gen)

<generator object my_range_iter at 0x000002BAB3C080B0>


In [80]:
next(my_gen)

StopIteration: 

In [81]:
for x in my_range_iter(10, 20):
    print('The value that is being yielded out as we loop over the generator is ', x)

The value that is being yielded out as we loop over the generator is  10
The value that is being yielded out as we loop over the generator is  11
The value that is being yielded out as we loop over the generator is  12
The value that is being yielded out as we loop over the generator is  13
The value that is being yielded out as we loop over the generator is  14
The value that is being yielded out as we loop over the generator is  15
The value that is being yielded out as we loop over the generator is  16
The value that is being yielded out as we loop over the generator is  17
The value that is being yielded out as we loop over the generator is  18
The value that is being yielded out as we loop over the generator is  19


In [82]:
for s in my_range_iter(10, 50, 5):
    print(s)

10
15
20
25
30
35
40
45


In [99]:
def read_file(filename):
    with open(filename) as large_file:
        for line in large_file:
            yield line
            
            
file = read_file('whiteboard.py')

In [125]:
next(file)

StopIteration: 

In [126]:
test_file = read_file('test.py')

for line in test_file:
    print(line)

from unittest import TestCase, main



from whiteboard import solution



class MatchTestCase(TestCase):

    def test_example_one(self):

        self.assertEqual(solution("test"), "e")

    def test_example_two(self):

        self.assertEqual(solution("teeter"), "r")

    def test_example_three(self):

        self.assertEqual(solution("trend"),"t" )

    def test_example_four(self):

        self.assertEqual(solution("aabbcc"),None)







if __name__ == '__main__':

    main()


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York', 'DC']`

In [128]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York", "DC"]


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [129]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [130]:
help(list.sort)

Help on method_descriptor:

sort(self, /, *, key=None, reverse=False)
    Sort the list in ascending order and return None.
    
    The sort is in-place (i.e. the list itself is modified) and stable (i.e. the
    order of two equal elements is maintained).
    
    If a key function is given, apply it once to each list item and sort them,
    ascending or descending, according to their function values.
    
    The reverse flag can be set to sort in descending order.



In [ ]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield", "David hassELHOFF", "Gary A.J. Bernstein"]


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]
`

In [ ]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angeles",44),("Miami",29)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

```
Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
```

In [135]:
def fib(num):
    pass

In [136]:
n = 6

for i in range(n):
    print(f"Iteration {i}: {fib(i)}")

Iteration 0: None
Iteration 1: None
Iteration 2: None
Iteration 3: None
Iteration 4: None
Iteration 5: None


### Exercise #5<br>
<p>Create a generator that takes a number argument and yields that number squared down to 0, then call the generator and iterate through and print the result.</p>

In [ ]:
    
# 100
# 81
# 64
# 49
# 36
# 25
# 16
# 9
# 4
# 1
# 0

In [ ]:
x = cool_generator(20)


In [ ]:
next(x) # 361